Fix(Reader): flavor with broken dates in history.xml

jusjusjus opened the pull request:
This is a hotfix to address the bug we encountered in some averaged MFFs with broken date entries.
https://github.com/BEL-CO/signal-service/issues/31

damian5710 reviewed (changes requested):
Looks good. Could you also bump the version and update the CHANGELOG.md file please?

ephathaway commented:
@jusjusjus could we do something like

```
@depricated('Use "mff_flavor" instead')
def flavor(self) -> str:
    """Old function"""

def mff_flavor(self) -> str:
    """New function"""
```

We use the `.flavor` method in MNE `read_evokeds_mff` function (https://github.com/mne-tools/mne-python/blob/6afc7224de9e7423e57604ca307941cf0b3449f1/mne/io/egi/egimff.py#L780) and require `mffpy>=0.5.7`.

jusjusjus commented:
@ephathaway, alright, I added package `Deprecated` and deprecated the function .l  Somehow I couldn't get it to through the dpereaction error.  Any experience with this?  Anyhow, I also added another test for the new property `Reader.mff_flavor`.

ephathaway commented on requirements.txt:
```diff
@@ -1,2 +1,3 @@
 pytz>=2019.2
 numpy>=1.15.1
+Deprecated>=1.2.12
```
Use `deprecated==1.2.0`. That worked for me. There must be some bug in the latest release.

jusjusjus replied:
Maybe we should include a decorator of our own device as long as module `deprecated`.

ephathaway replied:
That's also an option. This is my first time working with the `deprecated` package, so I am not super familiar with the functionality.

ephathaway replied:
It seems like v1.2.0 behaves like we want it to though.

jusjusjus replied:
Yeah, but that might lead to dependency clashes with downstream repositories.  Say, MNE wants to upgrade its version of module `deprecated` past v1.2.0.  This would clash with a strict requirement.

jusjusjus replied:
MNE [copied](https://github.com/mne-tools/mne-python/blob/2d1e3dfe569c569bf0331b7faca41c483e8b8237/mne/utils/docs.py#L2716) deprecated from scikit-learn into its code base.  If module `deprecated` matures, they might switch thus leading to clashes with our fixed dependency.

jusjusjus replied:
I think the behavior we find in module `deprecated` is actually correct.  The user of the application shouldn't see the warning, it's something that the developer should be aware of (see [here](https://www.lesinskis.com/python_deprecation_tutorial.html)).  So to check for the warnings one needs to run
```python
import warnings
warnings.simplefilter('default')
...
# raises warning
readerInstance.flavor
```

ephathaway replied:
Nice article. Yeah, that makes sense that the user wouldn't need or want to see the deprecation warning every time. I think as long as we make it clear in the `.flavor` docstring that it is deprecated, we will be good.

ephathaway reviewed (changes requested):
Let's make sure to increment the version to 0.6.3. See below about the deprecation

ephathaway commented on mffpy/reader.py:
````diff
@@ -61,13 +62,13 @@ def __init__(self, filename: str):
         ```python
         Reader.flavor
         ```
         return flavor of the MFF
````
Let's put a note to use `.mff_flavor` instead.

ephathaway reviewed (changes requested):
`numpy v1.20` requires `python>=3.7`. Let's either lock the `numpy` version in `requirements.txt` or specify `python 3.8` in the install instructions.

jusjusjus commented:
> `numpy v1.20` requires `python>=3.7`. Let's either lock the `numpy` version in `requirements.txt` or specify `python 3.8` in the install instructions.

But I don't think we are doing anything wrong specifying numpy>=1.15.1, then.  Numpy should have gone to version 2.

ephathaway commented:
> But I don't think we are doing anything wrong specifying numpy>=1.15.1, then. Numpy should have gone to version 2.

Gotcha. Yeah, I wasn't sure whether it's our responsibility to make sure all future versions of dependencies are compatible. In my personal opinion, it makes sense to me to lock dependency versions in general just so future releases don't break functionality. But I realize there's differing opinions on that.

jusjusjus commented:
> > But I don't think we are doing anything wrong specifying numpy>=1.15.1, then. Numpy should have gone to version 2.
> 
> Gotcha. Yeah, I wasn't sure whether it's our responsibility to make sure all future versions of dependencies are compatible. In my personal opinion, it makes sense to me to lock dependency versions in general just so future releases don't break functionality. But I realize there's differing opinions on that.
> 
> My main concern is that if I just directly follow the install steps in the readme, I get an error.

Actually, I wasn't able to reproduce the error.  I created new environments with py 3.6, 3.7, and 3.8 and it ran without problem.  Maybe there's something wrong with your environment, @ephathaway?

ephathaway commented:
> Actually, I wasn't able to reproduce the error. I created new environments with py 3.6, 3.7, and 3.8 and it ran without problem. Maybe there's something wrong with your environment, @ephathaway?

Hm... I also tested with a fresh environment on filadelfia. It was also a fresh clone of mffpy. Could it be the conda version? I've got 4.3.30.

jusjusjus commented:
> > Actually, I wasn't able to reproduce the error. I created new environments with py 3.6, 3.7, and 3.8 and it ran without problem. Maybe there's something wrong with your environment, @ephathaway?
> 
> Hm... I also tested with a fresh environment on filadelfia. It was also a fresh clone of mffpy. Could it be the conda version? I've got 4.3.30.
> 
> What version of numpy is automatically installed when you run `$ python setup.py install`?

Which python version do you mean?

damian5710 reviewed (commented):
Shouldn't we be using Python v3.8.8?

ephathaway commented:
> Which python version do you mean?

3.6

jusjusjus commented:
> > Which python version do you mean?
> 
> 3.6

Please find below my successful test with python 3.6, @ephathaway.  Can you spot the difference to your set up?
```
(base) jus@icedmocha:~/code/belco/mffpy$ conda create -n mffpy36 python=3.6
Collecting package metadata (repodata.json): done
Solving environment: done


==> WARNING: A newer version of conda exists. <==
  current version: 4.9.2
  latest version: 4.10.1

Please update conda by running

    $ conda update -n base -c defaults conda



## Package Plan ##

  environment location: /home/jus/anaconda3/envs/mffpy36

  added / updated specs:
    - python=3.6


The following NEW packages will be INSTALLED:

  _libgcc_mutex      pkgs/main/linux-64::_libgcc_mutex-0.1-main
  ca-certificates    pkgs/main/linux-64::ca-certificates-2021.4.13-h06a4308_1
  certifi            pkgs/main/linux-64::certifi-2020.12.5-py36h06a4308_0
  ld_impl_linux-64   pkgs/main/linux-64::ld_impl_linux-64-2.33.1-h53a641e_7
  libffi             pkgs/main/linux-64::libffi-3.3-he6710b0_2
  libgcc-ng          pkgs/main/linux-64::libgcc-ng-9.1.0-hdf63c60_0
  libstdcxx-ng       pkgs/main/linux-64::libstdcxx-ng-9.1.0-hdf63c60_0
  ncurses            pkgs/main/linux-64::ncurses-6.2-he6710b0_1
  openssl            pkgs/main/linux-64::openssl-1.1.1k-h27cfd23_0
  pip                pkgs/main/linux-64::pip-21.0.1-py36h06a4308_0
  python             pkgs/main/linux-64::python-3.6.13-hdb3f193_0
  readline           pkgs/main/linux-64::readline-8.1-h27cfd23_0
  setuptools         pkgs/main/linux-64::setuptools-52.0.0-py36h06a4308_0
  sqlite             pkgs/main/linux-64::sqlite-3.35.4-hdfb4753_0
  tk                 pkgs/main/linux-64::tk-8.6.10-hbc83047_0
  wheel              pkgs/main/noarch::wheel-0.36.2-pyhd3eb1b0_0
  xz                 pkgs/main/linux-64::xz-5.2.5-h7b6447c_0
  zlib               pkgs/main/linux-64::zlib-1.2.11-h7b6447c_3


Proceed ([y]/n)? y

Preparing transaction: done
Verifying transaction: done
Executing transaction: done
#
# To activate this environment, use
#
#     $ conda activate mffpy36
#
# To deactivate an active environment, use
#
#     $ conda deactivate

(base) jus@icedmocha:~/code/belco/mffpy$ activate mffpy36
(mffpy36) jus@icedmocha:~/code/belco/mffpy$ pip install -r requirements.txt -r requirements-dev.txt 
Collecting pytz>=2019.2
  Using cached pytz-2021.1-py2.py3-none-any.whl (510 kB)
Collecting numpy>=1.15.1
  Using cached numpy-1.19.5-cp36-cp36m-manylinux2010_x86_64.whl (14.8 MB)
Collecting Deprecated>=1.2.12
  Using cached Deprecated-1.2.12-py2.py3-none-any.whl (9.5 kB)
Collecting mypy>=0.660
  Using cached mypy-0.812-cp36-cp36m-manylinux2010_x86_64.whl (21.6 MB)
Collecting pytest>=5.2.0
  Using cached pytest-6.2.4-py3-none-any.whl (280 kB)
Collecting pytest-cov>=2.6.1
  Using cached pytest_cov-2.12.0-py2.py3-none-any.whl (20 kB)
Collecting pre-commit>=2.7.1
  Using cached pre_commit-2.12.1-py2.py3-none-any.whl (189 kB)
Collecting flake8==3.8.4
  Using cached flake8-3.8.4-py2.py3-none-any.whl (72 kB)
Collecting importlib-metadata
  Using cached importlib_metadata-4.0.1-py3-none-any.whl (16 kB)
Collecting pyflakes<2.3.0,>=2.2.0
  Using cached pyflakes-2.2.0-py2.py3-none-any.whl (66 kB)
Collecting mccabe<0.7.0,>=0.6.0
  Using cached mccabe-0.6.1-py2.py3-none-any.whl (8.6 kB)
Collecting pycodestyle<2.7.0,>=2.6.0a1
  Using cached pycodestyle-2.6.0-py2.py3-none-any.whl (41 kB)
Collecting wrapt<2,>=1.10
  Using cached wrapt-1.12.1-cp36-cp36m-linux_x86_64.whl
Collecting typing-extensions>=3.7.4
  Using cached typing_extensions-3.10.0.0-py3-none-any.whl (26 kB)
Collecting mypy-extensions<0.5.0,>=0.4.3
  Using cached mypy_extensions-0.4.3-py2.py3-none-any.whl (4.5 kB)
Collecting typed-ast<1.5.0,>=1.4.0
  Using cached typed_ast-1.4.3-cp36-cp36m-manylinux1_x86_64.whl (743 kB)
Collecting nodeenv>=0.11.1
  Using cached nodeenv-1.6.0-py2.py3-none-any.whl (21 kB)
Collecting virtualenv>=20.0.8
  Using cached virtualenv-20.4.6-py2.py3-none-any.whl (7.2 MB)
Collecting toml
  Using cached toml-0.10.2-py2.py3-none-any.whl (16 kB)
Collecting importlib-resources
  Using cached importlib_resources-5.1.3-py3-none-any.whl (26 kB)
Collecting pyyaml>=5.1
  Using cached PyYAML-5.4.1-cp36-cp36m-manylinux1_x86_64.whl (640 kB)
Collecting cfgv>=2.0.0
  Using cached cfgv-3.3.0-py2.py3-none-any.whl (7.3 kB)
Collecting identify>=1.0.0
  Using cached identify-2.2.4-py2.py3-none-any.whl (98 kB)
Collecting packaging
  Using cached packaging-20.9-py2.py3-none-any.whl (40 kB)
Collecting pluggy<1.0.0a1,>=0.12
  Using cached pluggy-0.13.1-py2.py3-none-any.whl (18 kB)
Collecting attrs>=19.2.0
  Using cached attrs-21.2.0-py2.py3-none-any.whl (53 kB)
Collecting py>=1.8.2
  Using cached py-1.10.0-py2.py3-none-any.whl (97 kB)
Collecting iniconfig
  Using cached iniconfig-1.1.1-py2.py3-none-any.whl (5.0 kB)
Collecting coverage[toml]>=5.2.1
  Using cached coverage-5.5-cp36-cp36m-manylinux2010_x86_64.whl (240 kB)
Collecting zipp>=0.5
  Using cached zipp-3.4.1-py3-none-any.whl (5.2 kB)
Collecting appdirs<2,>=1.4.3
  Using cached appdirs-1.4.4-py2.py3-none-any.whl (9.6 kB)
Collecting distlib<1,>=0.3.1
  Using cached distlib-0.3.1-py2.py3-none-any.whl (335 kB)
Collecting six<2,>=1.9.0
  Using cached six-1.16.0-py2.py3-none-any.whl (11 kB)
Collecting filelock<4,>=3.0.0
  Using cached filelock-3.0.12-py3-none-any.whl (7.6 kB)
Collecting pyparsing>=2.0.2
  Using cached pyparsing-2.4.7-py2.py3-none-any.whl (67 kB)
Installing collected packages: zipp, typing-extensions, pyparsing, importlib-metadata, toml, six, py, pluggy, packaging, iniconfig, importlib-resources, filelock, distlib, coverage, attrs, appdirs, wrapt, virtualenv, typed-ast, pyyaml, pytest, pyflakes, pycodestyle, nodeenv, mypy-extensions, mccabe, identify, cfgv, pytz, pytest-cov, pre-commit, numpy, mypy, flake8, Deprecated
Successfully installed Deprecated-1.2.12 appdirs-1.4.4 attrs-21.2.0 cfgv-3.3.0 coverage-5.5 distlib-0.3.1 filelock-3.0.12 flake8-3.8.4 identify-2.2.4 importlib-metadata-4.0.1 importlib-resources-5.1.3 iniconfig-1.1.1 mccabe-0.6.1 mypy-0.812 mypy-extensions-0.4.3 nodeenv-1.6.0 numpy-1.19.5 packaging-20.9 pluggy-0.13.1 pre-commit-2.12.1 py-1.10.0 pycodestyle-2.6.0 pyflakes-2.2.0 pyparsing-2.4.7 pytest-6.2.4 pytest-cov-2.12.0 pytz-2021.1 pyyaml-5.4.1 six-1.16.0 toml-0.10.2 typed-ast-1.4.3 typing-extensions-3.10.0.0 virtualenv-20.4.6 wrapt-1.12.1 zipp-3.4.1
(mffpy36) jus@icedmocha:~/code/belco/mffpy$ make test 
# zip -Z store -r -j ./examples/example_1.mfz ./examples/example_1.mff
python ./bin/mff2mfz.py ./examples/example_1.mff
----------------------------------------------------------
Storing the following files in './examples/example_1.mfz':
----------------------------------------------------------
> ./examples/example_1.mff/coordinates.xml
> ./examples/example_1.mff/signal1.bin
> ./examples/example_1.mff/info.xml
> ./examples/example_1.mff/epochs.xml
> ./examples/example_1.mff/subject.xml
> ./examples/example_1.mff/sensorLayout.xml
> ./examples/example_1.mff/dipoleSet.xml
> ./examples/example_1.mff/categories.xml
> ./examples/example_1.mff/Events_ECI.xml
> ./examples/example_1.mff/info1.xml
----------------------------------------------------------
mypy --ignore-missing-imports mffpy
Success: no issues found in 29 source files
pytest --cov
========================================== test session starts ===========================================
platform linux -- Python 3.6.13, pytest-6.2.4, py-1.10.0, pluggy-0.13.1
rootdir: /home/jus/code/belco/mffpy
plugins: cov-2.12.0
collected 109 items                                                                                      

mffpy/tests/test_cached_property.py ..                                                             [  1%]
mffpy/tests/test_devices.py ...........                                                            [ 11%]
mffpy/tests/test_dict2xml.py .                                                                     [ 12%]
mffpy/tests/test_header_block.py ..                                                                [ 14%]
mffpy/tests/test_mffdir.py ....                                                                    [ 18%]
mffpy/tests/test_raw_bin_files.py ..............                                                   [ 31%]
mffpy/tests/test_reader.py ......................                                                  [ 51%]
mffpy/tests/test_writer.py .........                                                               [ 59%]
mffpy/tests/test_xml_files.py .......................................                              [ 95%]
mffpy/tests/test_zipfile.py .....                                                                  [100%]

============================================ warnings summary ============================================
mffpy/tests/test_reader.py::test_flavor
mffpy/tests/test_reader.py::test_flavor
mffpy/tests/test_reader.py::test_flavor
mffpy/tests/test_reader.py::test_flavor
  /home/jus/code/belco/mffpy/mffpy/cached_property.py:36: DeprecationWarning: Call to deprecated function (or staticmethod) flavor. (Use ".mff_flavor" instead) -- Deprecated since version 0.6.3.
    ans = fn(self)

-- Docs: https://docs.pytest.org/en/stable/warnings.html

---------- coverage: platform linux, python 3.6.13-final-0 -----------
Name                                          Stmts   Miss  Cover
-----------------------------------------------------------------
mffpy/__init__.py                                 4      0   100%
mffpy/bin_files.py                               40      2    95%
mffpy/bin_writer.py                              71      0   100%
mffpy/cached_property.py                         25      1    96%
mffpy/devices.py                                 10      0   100%
mffpy/dict2xml.py                                31      3    90%
mffpy/epoch.py                                   24      3    88%
mffpy/header_block/__init__.py                    1      0   100%
mffpy/header_block/header_block.py               48      2    96%
mffpy/header_block/helpers.py                    15      0   100%
mffpy/header_block/optional_header_block.py      32      1    97%
mffpy/mffdir.py                                  92      7    92%
mffpy/raw_bin_files.py                           99      0   100%
mffpy/reader.py                                 110      2    98%
mffpy/tests/__init__.py                           0      0   100%
mffpy/tests/test_cached_property.py              33      0   100%
mffpy/tests/test_devices.py                      12      0   100%
mffpy/tests/test_dict2xml.py                     15      0   100%
mffpy/tests/test_header_block.py                 33      0   100%
mffpy/tests/test_mffdir.py                       30      0   100%
mffpy/tests/test_raw_bin_files.py                37      0   100%
mffpy/tests/test_reader.py                       95      0   100%
mffpy/tests/test_writer.py                      176      8    95%
mffpy/tests/test_xml_files.py                   185      1    99%
mffpy/tests/test_zipfile.py                      34      0   100%
mffpy/version.py                                  1      0   100%
mffpy/writer.py                                  63      2    97%
mffpy/xml_files.py                              542     18    97%
mffpy/zipfile.py                                 47      0   100%
-----------------------------------------------------------------
TOTAL                                          1905     50    97%

==================================== 109 passed, 4 warnings in 4.71s =====================================
```

ephathaway commented:
@jusjusjus (conda 4.10.1)
```
(base) evan@filadelfia:~/Repositories/mffpy$ conda create -n mffpy36 python=3.6
(base) evan@filadelfia:~/Repositories/mffpy$ conda activate mffpy36
(mffpy36) evan@filadelfia:~/Repositories/mffpy$ pip install -r requirements.txt -r requirements-dev.txt
(mffpy36) evan@filadelfia:~/Repositories/mffpy$ make test
```
`numpy-1.19.5` is installed, no errors.
```
# Following install instructions in README
(base) evan@filadelfia:~/Repositories/mffpy$ conda create -n mffpy36 python=3.6 pip
(base) evan@filadelfia:~/Repositories/mffpy$ conda activate mffpy36
(mffpy36) evan@filadelfia:~/Repositories/mffpy$ pip install -r requirements-dev.txt
(mffpy36) evan@filadelfia:~/Repositories/mffpy$ python setup.py install
...
Searching for numpy>=1.15.1
Reading https://pypi.org/simple/numpy/
Downloading https://files.pythonhosted.org/packages/f3/1f/fe9459e39335e7d0e372b5e5dcd60f4381d3d1b42f0b9c8222102ff29ded/numpy-1.20.3.zip#sha256=e55185e51b18d788e49fe8305fd73ef4470596b33fc2c1ceb304566b99c71a69
Best match: numpy 1.20.3
Processing numpy-1.20.3.zip
Writing /tmp/easy_install-jgia7ymh/numpy-1.20.3/setup.cfg
Running numpy-1.20.3/setup.py -q bdist_egg --dist-dir /tmp/easy_install-jgia7ymh/numpy-1.20.3/egg-dist-tmp-eh5yumc4
Traceback (most recent call last):
  File "/home/evan/anaconda3/envs/mffpy36/lib/python3.6/site-packages/setuptools/sandbox.py", line 152, in save_modules
    yield saved
  File "/home/evan/anaconda3/envs/mffpy36/lib/python3.6/site-packages/setuptools/sandbox.py", line 193, in setup_context
    yield
  File "/home/evan/anaconda3/envs/mffpy36/lib/python3.6/site-packages/setuptools/sandbox.py", line 254, in run_setup
    _execfile(setup_script, ns)
  File "/home/evan/anaconda3/envs/mffpy36/lib/python3.6/site-packages/setuptools/sandbox.py", line 43, in _execfile
    exec(code, globals, locals)
  File "/tmp/easy_install-jgia7ymh/numpy-1.20.3/setup.py", line 30, in <module>
    "License :: OSI Approved :: Apache Software License",
RuntimeError: Python version >= 3.7 required.
```

jusjusjus commented:
> @jusjusjus (conda 4.10.1)
> 
> ```
> (base) evan@filadelfia:~/Repositories/mffpy$ conda create -n mffpy36 python=3.6
> (base) evan@filadelfia:~/Repositories/mffpy$ conda activate mffpy36
> (mffpy36) evan@filadelfia:~/Repositories/mffpy$ pip install -r requirements.txt -r requirements-dev.txt
> (mffpy36) evan@filadelfia:~/Repositories/mffpy$ make test
> ```
> 
> `numpy-1.19.5` is installed, no errors.
> 
> ```
> # Following install instructions in README
> (base) evan@filadelfia:~/Repositories/mffpy$ conda create -n mffpy36 python=3.6 pip
> (base) evan@filadelfia:~/Repositories/mffpy$ conda activate mffpy36
> (mffpy36) evan@filadelfia:~/Repositories/mffpy$ pip install -r requirements-dev.txt
> (mffpy36) evan@filadelfia:~/Repositories/mffpy$ python setup.py install
> ...
> Searching for numpy>=1.15.1
> Reading https://pypi.org/simple/numpy/
> Downloading https://files.pythonhosted.org/packages/f3/1f/fe9459e39335e7d0e372b5e5dcd60f4381d3d1b42f0b9c8222102ff29ded/numpy-1.20.3.zip#sha256=e55185e51b18d788e49fe8305fd73ef4470596b33fc2c1ceb304566b99c71a69
> Best match: numpy 1.20.3
> Processing numpy-1.20.3.zip
> Writing /tmp/easy_install-jgia7ymh/numpy-1.20.3/setup.cfg
> Running numpy-1.20.3/setup.py -q bdist_egg --dist-dir /tmp/easy_install-jgia7ymh/numpy-1.20.3/egg-dist-tmp-eh5yumc4
> Traceback (most recent call last):
>   File "/home/evan/anaconda3/envs/mffpy36/lib/python3.6/site-packages/setuptools/sandbox.py", line 152, in save_modules
>     yield saved
>   File "/home/evan/anaconda3/envs/mffpy36/lib/python3.6/site-packages/setuptools/sandbox.py", line 193, in setup_context
>     yield
>   File "/home/evan/anaconda3/envs/mffpy36/lib/python3.6/site-packages/setuptools/sandbox.py", line 254, in run_setup
>     _execfile(setup_script, ns)
>   File "/home/evan/anaconda3/envs/mffpy36/lib/python3.6/site-packages/setuptools/sandbox.py", line 43, in _execfile
>     exec(code, globals, locals)
>   File "/tmp/easy_install-jgia7ymh/numpy-1.20.3/setup.py", line 30, in <module>
>     "License :: OSI Approved :: Apache Software License",
> RuntimeError: Python version >= 3.7 required.
> ```
> 
> For some reason, allowing the `setup.py` script to install the dependencies will try to install `numpy-1.20.3`.

Yeah, something is broken in when executing setup.py in the raw.  Technically, pip also reads setup.py, but it must use a different machine to install the dependencies.  There is not released version for numpy 1.20 for py36, so It's very curious that the resolution process fails in this way.  Good that you caught that.

jusjusjus commented:
> Shouldn't we be using Python v3.8.8?

@damian5710, while our software uses a more up-to-date version of python, this public package supports usage of python 3.6 and up because it is used in MNE-python which is python>=3.7.  We can still support 3.6 as well.
